### ISSUE TITLE
server responding with incorrectly encoded redirects breaks in ktor-client

### ISSUE THREAD

**Issue by NikkyAI** [ktorio/ktor#607]
this works in chrome, wget, curl

sample url: 
```
https://files.forgecdn.net/files/2574/880/BiblioCraft[v2.4.5][MC1.12.2].jar
https://files.forgecdn.net/files/2611/560/Botania r1.10-356.jar
https://files.forgecdn.net/files/2613/730/Toast Control-1.12.2-1.7.1.jar
```

as far as i can tell the characters that are missed in encoding are `%20`, `%2B`, `%2D`
but there is probably more to it

the Apache engine itself will barf at me for the redirects, and handling them myself in a HttRedirect featue just gets me a error xml, even when just requesting the last url in the redirect chain (the same url work fine in chrome)

**Comment by NikkyAI** [comment]
my currently working implementation, not using ktor and quite ugly: 
```
suspend fun File.download(
    url: String,
    cacheDir: File,
    logger: KLogger = Downloader.logger
) {
    val cacheFile = cacheDir.resolve(this.name)
    logger.info("downloading $url -> ${this@download}")
    logger.debug("cacheFile $cacheFile")
    if (cacheFile.exists() && !cacheFile.isFile) cacheFile.deleteRecursively()

    if (!cacheFile.exists() || !cacheFile.isFile) {
        var nextUrl = url.encoded
        do {
            nextUrl = nextUrl.encoded
            logger.info { nextUrl }
            val (request, response, result) = nextUrl //.encode()
                .httpGet().header("User-Agent" to Downloader.useragent)
                .allowRedirects(false)
                .awaitByteArrayResponse()
            val isRedirect = when (result) {
                is Result.Success -> {
                    cacheDir.mkdirs()
                    cacheFile.parentFile.mkdirs()
                    cacheFile.writeBytes(result.value)
                    false
                }
                is Result.Failure -> {
                    if (response.isStatusRedirection) {
                        nextUrl = response.headers["Location"]?.firstOrNull() ?:
                            throw IllegalStateException("missing Location header")
                        true
                    } else {
                        logger.error("invalid statusCode {} from {}", response.statusCode, url.encoded)
                        logger.error("connection url: {}", request.url)
                        logger.error("content: {}", result.component1())
                        logger.error("error: {}", result.error.toString())
                        throw IOException(result.error.toString())
                    }
                }
            }
        } while (isRedirect)
    }

    logger.debug("saving $url -> $this")
    try {
        this.parentFile.mkdirs()
        cacheFile.copyTo(this, overwrite = true)
    } catch (e: FileAlreadyExistsException) {
        val fileIsLocked = !this.renameTo(this)
        logger.error("failed to copy file $cacheFile to $this .. file is locked ? $fileIsLocked")
        if (!fileIsLocked)
            cacheFile.copyTo(this, overwrite = true)
    }

    logger.debug("done downloading $url -> $this")
}

val String.encoded: String
    get() = this
        .replace(" ", "%20")
        .replace("[", "%5b")
        .replace("]", "%5d")
```

**Comment by e5l** [comment]
Hi, @NikkyAI. Thanks for the report.
It's fixed in master and `1.0.0-alpha-1`
